File: interviews/lc/04-contains-duplicate.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>
#include "lc-utility.h"

using std::string;
using std::unordered_set;
using std::vector;

bool containsDuplicate(const vector<int>& nums) {
  unordered_set<int> seen;
  for (const auto n : nums) {
    if (seen.count(n) > 0) {
      return true;
    }
    seen.insert(n);
  }
  return false;
}

void runTestCase(const vector<int> nums, bool expected) {
  const auto numsStr = vector2str(nums);
  const auto actual = containsDuplicate(nums);
  std::cout << numsStr << " => Expected: " << std::to_string(expected)
            << " Actual: " << std::to_string(actual) << std::endl;
}

int main(int argc, char* argv[]) {
  runTestCase({1, 2, 3, 1}, true);
  runTestCase({1, 2, 3, 4}, false);
  runTestCase({1, 1, 1, 3, 3, 4, 3, 2, 4, 2}, true);
  return EXIT_SUCCESS;
}

File: interviews/lc/04-contains-duplicate.py
from typing import List


def containsDuplicate(nums: List[int]) -> bool:
    seen = {}
    for i in nums:
        if i in seen:
            return True
        seen[i] = True
    return False


def runTestCase(nums: List[int], expected: bool) -> None:
    numsStr = ",".join([str(n) for n in nums])
    actual = containsDuplicate(nums)
    print(f"{numsStr} => Expected: {expected} Actual: {actual}")


if __name__ == "__main__":
    runTestCase([1, 2, 3, 1], True)
    runTestCase([1, 2, 3, 4], False)
    runTestCase([1, 1, 1, 3, 3, 4, 3, 2, 4, 2], True)
